vault/resource_ldap_secret_backend.go

```diff
@@ -0,0 +1,530 @@
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: `Maximum TLS version to use. Accepted values are 'tls10', 'tls11', 'tls12' or 'tls13'. Defaults to 'tls12'`,
+		},
+		"tls_min_version": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: `Minimum TLS version to use. Accepted values are 'tls10', 'tls11', 'tls12' or 'tls13'. Defaults to 'tls12'`,
+		},
+		consts.FieldTTL: {
+			Type:        schema.TypeInt,
+			Optional:    true,
+			Computed:    true,
+			Description: `In seconds, the default password time-to-live.`,
+		},
+		"upndomain": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: `Enables userPrincipalDomain login with [username]@UPNDomain.`,
+		},
+		"url": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: `LDAP URL to connect to (default: ldap://127.0.0.1). Multiple URLs can be specified by concatenating them with commas; they will be tried in-order.`,
+		},
+		"userattr": {
+			Type:        schema.TypeString,
+			Default:     "cn",
+			Optional:    true,
+			Description: `Attribute used for users (default: cn)`,
+		},
+		"userdn": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: `LDAP domain to use for users (eg: ou=People,dc=example,dc=org)`,
+		},
+	}
+	return provider.MustAddMountMigrationSchema(&schema.Resource{
+		CreateContext: createLDAPConfigResource,
+		UpdateContext: updateLDAPConfigResource,
+		ReadContext:   ReadContextWrapper(readLDAPConfigResource),
+		DeleteContext: deleteLDAPConfigResource,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		CustomizeDiff: getMountCustomizeDiffFunc(consts.FieldBackend),
+		Schema:        fields,
+	})
+}
+
+func createLDAPConfigResource(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, err := provider.GetClient(d, meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	backend := d.Get(consts.FieldBackend).(string)
+	description := d.Get("description").(string)
+	defaultTTL := d.Get(consts.FieldDefaultLeaseTTL).(int)
+	maxTTL := d.Get(consts.FieldMaxLeaseTTL).(int)
+
+	log.Printf("[DEBUG] Mounting LDAP backend at %q", backend)
+	err = client.Sys().Mount(backend, &api.MountInput{
+		Type:        consts.MountTypeLDAP,
+		Description: description,
+		Config: api.MountConfigInput{
+			DefaultLeaseTTL: fmt.Sprintf("%ds", defaultTTL),
+			MaxLeaseTTL:     fmt.Sprintf("%ds", maxTTL),
+		},
+	})
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error mounting to %q: %s", backend, err))
+	}
+
+	log.Printf("[DEBUG] Mounted LDAP backend at %q", backend)
+	d.SetId(backend)
+
+	data := map[string]interface{}{}
+	if v, ok := d.GetOk("binddn"); ok {
+		data["binddn"] = v
+	}
+	if v, ok := d.GetOk("bindpass"); ok {
+		data["bindpass"] = v
+	}
+	if v, ok := d.GetOk("case_sensitive_names"); ok {
```

Comment:
I believe the recommendation is to use `d.Get` for booleans because we can't really trust that the exists response is accurate.

Reply:
can you elaborate on this? It looks like `d.Get` just defers to `d.GetOk`. If the key is missing, then we don't set the `data` map.

Reply:
For most boolean fields, using the Get() method is preferred over GetOk(). The GetOk() evaluates the boolean and returns its value, as opposed to returning whether the boolean was set or not. This does not work in the scenario that the boolean’s value is false, as d.GetOk evaluates the boolean value and returns an ok value of false , thinking that the boolean was not set. But in fact, it was set, just with a value of false.

Reply:
Oh strange behavior.. I updated the test to account for this. Thanks! I wouldn't have caught that if you didn't point it out